File: src/forager/ui/workers.py
"""Background QThreads and job functions for scanning and art fetching.

Network / disk work never runs on the GUI thread; these workers emit results
back through Qt signals. Each exposes a stop ``threading.Event`` so the
window can cancel cleanly on shutdown.
"""
from __future__ import annotations
import threading
from PySide6.QtCore import QThread, QObject, Signal

from forager.core.game import Game
from forager.library.scanner import scan_all


class ScanWorker(QThread):
    done = Signal(object)

    def run(self):
        games = scan_all()
        if not self.isInterruptionRequested():
            self.done.emit(games)


class ProtonUpdateWorker(QThread):
    message = Signal(str)
    progress = Signal(object)
    done = Signal(bool, str)

    def __init__(self, cancel_event: threading.Event | None = None, parent=None):
        super().__init__(parent)
        self._cancel = cancel_event

    def run(self):
        from forager.compatibility.proton import update_proton, DownloadCancelled

        try:
            version = update_proton(
                self.message.emit,
                self.progress.emit,
                self._cancel,
            )
        except DownloadCancelled:
            self.done.emit(False, "Download cancelled")
        except Exception as e:
            self.done.emit(False, str(e))
        else:
            self.done.emit(True, version or "")


class ToolUpdateSignals(QObject):
    done = Signal(list)


def _tool_update_check_job(signals: ToolUpdateSignals, stop_event: threading.Event):
    from forager.updates.tool_updates import check_tool_updates

    if stop_event.is_set():
        return
    try:
        updates = check_tool_updates()
    except Exception:
        updates = []
    if not stop_event.is_set():
        signals.done.emit(updates)


class ToolUpdateWorker(QThread):
    done = Signal(bool, str)

    def run(self):
        from forager.updates.tool_updates import update_tool_updates

        try:
            updated = update_tool_updates()
        except Exception as e:
            self.done.emit(False, str(e))
        else:
            if updated:
                self.done.emit(True, "Updated: " + ", ".join(updated))
            else:
                self.done.emit(True, "Tools are up to date")


class ArtSignals(QObject):
    grid_ready = Signal(object)
    icon_ready = Signal(object)


class HeroSignals(QObject):
    ready = Signal(object)


def _art_job(games: list[Game], signals: ArtSignals, stop_event: threading.Event):
    from forager.artwork import pipeline as art
    from forager.services.icon_provider import load_icon_bytes

    for game in games:
        if stop_event.is_set():
            return
        data = art.load_grid_bytes(game)
        if data:
            signals.grid_ready.emit((game, data))
        if stop_event.is_set():
            return
        icon = load_icon_bytes(game)
        if icon:
            signals.icon_ready.emit((game, icon))


def _hero_job(game: Game, signals: HeroSignals, stop_event: threading.Event):
    from forager.artwork import pipeline as art

    if stop_event.is_set():
        return
    data = art.load_hero_bytes(game)
    if data:
        signals.ready.emit((game, data))

File: src/ui/Workers.cpp
#include "ui/Workers.h"
#include "library/Scanner.h"
#include "artwork/Pipeline.h"
#include "artwork/PixmapUtils.h"
#include <QMutexLocker>
#include <QBuffer>
#include <algorithm>

// -- ScanWorker ------------------------------------------------------------

ScanWorker::ScanWorker(QObject* parent)
    : QThread(parent)
{
}

void ScanWorker::run() {
    std::vector<Game> vec = scanner::scanAll();
    QList<Game> games;
    games.reserve(vec.size());
    for (auto& g : vec) games.append(std::move(g));
    if (!isInterruptionRequested())
        emit done(games);
}

// -- ProtonUpdateWorker ----------------------------------------------------

ProtonUpdateWorker::ProtonUpdateWorker(QObject* parent)
    : QThread(parent)
{
}

void ProtonUpdateWorker::run() {
    emit done(false, "Proton update not yet implemented");
}

// -- DownloadWorker --------------------------------------------------------

DownloadWorker::DownloadWorker(const QString& providerName, const QString& appId,
                               const QString& destination, QObject* parent)
    : QThread(parent), m_provider(providerName), m_appId(appId), m_dest(destination)
{
}

void DownloadWorker::cancel() {
    QMutexLocker lock(&m_mutex);
    m_cancelled = true;
}

void DownloadWorker::run() {
    emit done(false, "Download not yet implemented");
}

// -- Art job functions -----------------------------------------------------

void workers::runArtJob(const QList<Game>& games, ArtSignals* artSignals, std::function<bool()> isStopped) {
    for (const auto& game : games) {
        if (isStopped()) return;
        QPixmap grid = art::loadGrid(game, true);
        if (!grid.isNull())
            emit artSignals->gridReady(game, pixmap::toJpeg(grid));
        if (isStopped()) return;
        QPixmap icon = art::loadIcon(game, true);
        if (!icon.isNull()) {
            QByteArray png;
            QBuffer buf(&png);
            buf.open(QIODevice::WriteOnly);
            if (icon.save(&buf, "PNG"))
                emit artSignals->iconReady(game, png);
        }
        if (isStopped()) return;
    }
}

void workers::runHeroJob(const Game& game, HeroSignals* heroSignals, std::function<bool()> isStopped) {
    if (isStopped()) return;
    QPixmap pix = art::loadHero(game, true);
    if (!pix.isNull())
        emit heroSignals->ready(game, pixmap::toJpeg(pix));
}
